src/js/background.ts

```diff
@@ -135,67 +162,108 @@ function handleMessages(
+        sendResponse({valid: false, reason: 'no matching origin'});
+        return;
+      }
+      const manifestObj = origin.get(message.version);
+      const manifest = manifestObj && manifestObj.leaves;
+      if (!manifest) {
+        sendResponse({valid: false, reason: 'no matching manifest'});
+        return;
+      }
+
+      // fetch the src
+      const encoder = new TextEncoder();
+      const encodedJS = encoder.encode(message.rawjs);
+      // hash the src
+      crypto.subtle.digest('SHA-256', encodedJS).then(jsHashBuffer => {
+        const jsHashArray = Array.from(new Uint8Array(jsHashBuffer));
+        const jsHash = jsHashArray
+          .map(b => b.toString(16).padStart(2, '0'))
+          .join('');
+
+        if (manifestObj.leaves.includes(jsHash)) {
+          sendResponse({valid: true, hash: jsHash});
+        } else {
+          sendResponse({
+            valid: false,
+            hash: jsHash,
+            reason:
+              'Error: hash does not match ' +
+              message.origin +
+              ', ' +
+              message.version +
+              ', unmatched JS is ' +
+              message.rawjs,
+          });
+        }
+      });
+
+      // Indicates that the message will send an async response.
+      return true;
     }
-    const manifestObj = origin.get(message.version);
-    const manifest = manifestObj && manifestObj.leaves;
-    if (!manifest) {
-      sendResponse({valid: false, reason: 'no matching manifest'});
+
+    case MESSAGE_TYPE.UPDATE_STATE: {
+      updateContentScriptState(validSender, message.state, message.origin);
+      sendResponse({success: true});
       return;
     }
 
-    // fetch the src
-    const encoder = new TextEncoder();
-    const encodedJS = encoder.encode(message.rawjs);
-    // hash the src
-    crypto.subtle.digest('SHA-256', encodedJS).then(jsHashBuffer => {
-      const jsHashArray = Array.from(new Uint8Array(jsHashBuffer));
-      const jsHash = jsHashArray
-        .map(b => b.toString(16).padStart(2, '0'))
-        .join('');
-
-      if (manifestObj.leaves.includes(jsHash)) {
-        sendResponse({valid: true});
-      } else {
+    case MESSAGE_TYPE.CONTENT_SCRIPT_START: {
+      recordContentScriptStart(validSender, message.origin);
+
+      if (!message.checkCSPHeaders) {
         sendResponse({
-          valid: false,
-          hash: jsHash,
-          reason:
-            'Error: hash does not match ' +
-            message.origin +
-            ', ' +
-            message.version +
-            ', unmatched JS is ' +
-            message.rawjs,
+          success: true,
         });
+
+        return;
+      }
+
+      const cspHeadersByTab = CSP_HEADERS.get(validSender.tab.id);
+      if (!cspHeadersByTab) {
+        // This may indicate that setupCSPListener has failed to fire.
+        invalidateContentScriptStateAndThrow(
+          validSender,
+          message.origin,
+          'No CSP headers have been stored for this tab ID',
```

Comment:
Nice, took me a long time to debug this last time; great that we now have an explicit check for it.